File: src/main/java/io/vertx/ext/sql/ResultSet.java
package io.vertx.ext.sql;

import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents the results of a SQL query.
 * <p>
 * It contains a list for the column names of the results, and a list of {@code JsonArray} - one for each row of the
 * results.
 *
 * @author <a href="http://tfox.org">Tim Fox</a>
 */
@DataObject(generateConverter = true)
public class ResultSet {

  private List<String> columnNames;
  private List<JsonArray> results;
  private List<JsonObject> rows;
  private JsonArray output;
  // chainned result sets
  private ResultSet next;

  /**
   * Default constructor
   */
  public ResultSet() {
  }

  /**
   * Copy constructor
   *
   * @param other  result-set to copy
   */
  public ResultSet(ResultSet other) {
    this.columnNames = other.columnNames;
    this.results = other.results;
    this.output = other.output;
    this.next = other.next;
  }

  /**
   * Create a result-set
   *
   * @param columnNames  the column names
   * @param results  the results
   */
  public ResultSet(List<String> columnNames, List<JsonArray> results, ResultSet next) {
    this.columnNames = columnNames;
    this.results = results;
    this.next = next;
  }

  /**
   * Create a result-set from JSON
   *
   * @param json  the json
   */
  @SuppressWarnings("unchecked")
  public ResultSet(JsonObject json) {
    ResultSetConverter.fromJson(json, this);
  }

  /**
   * Convert to JSON
   *
   * @return json object
   */
  public JsonObject toJson() {
    JsonObject obj = new JsonObject();
    ResultSetConverter.toJson(this, obj);
    return obj;
  }

  /**
   * Get the results
   *
   * @return the results
   */
  public List<JsonArray> getResults() {
    return results;
  }

  public ResultSet setResults(List<JsonArray> results) {
    this.results = results;
    return this;
  }

  /**
   * Get the registered outputs
   *
   * @return the outputs
   */
  public JsonArray getOutput() {
    return output;
  }

  public ResultSet setOutput(JsonArray output) {
    this.output = output;
    return this;
  }

  /**
   * Get the column names
   *
   * @return the column names
   */
  public List<String> getColumnNames() {
    return columnNames;
  }

  public ResultSet setColumnNames(List<String> columnNames) {
    this.columnNames = columnNames;
    return this;
  }

  /**
   * Get the next result set
   *
   * @return the next resultset
   */
  public ResultSet getNext() {
    return next;
  }

  public ResultSet setNext(ResultSet next) {
    this.next = next;
    return this;
  }

  /**
   * Get the rows - each row represented as a JsonObject where the keys are the column names and the values are
   * the column values.
   * <p>
   * Beware that it's legal for a query result in SQL to contain duplicate column names, in which case one will
   * overwrite the other if using this method. If that's the case use {@link #getResults} instead.
   *
   * @return  the rows represented as JSON object instances
   */
  public List<JsonObject> getRows() {
    if (rows == null) {
      rows = new ArrayList<>(results.size());
      int cols = columnNames.size();
      for (JsonArray result: results) {
        JsonObject row = new JsonObject();
        for (int i = 0; i < cols; i++) {
          row.put(columnNames.get(i), result.getValue(i));
        }
        rows.add(row);
      }
    }
    return rows;
  }

  /**
   * Return the number of rows in the result set
   *
   * @return the number of rows
   */
  public int getNumRows() {
    return results.size();
  }

  /**
   * Return the number of columns in the result set
   *
   * @return the number of columns
   */
  public int getNumColumns() {
    return columnNames.size();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    ResultSet resultSet = (ResultSet) o;

    if (columnNames != null ? !columnNames.equals(resultSet.columnNames) : resultSet.columnNames != null) return false;
    if (results != null ? !results.equals(resultSet.results) : resultSet.results != null) return false;
    if (next != null ? !next.equals(resultSet.next) : resultSet.next != null) return false;

    return true;
  }

  @Override
  public int hashCode() {
    int result = columnNames != null ? columnNames.hashCode() : 0;
    result = 31 * result + (results != null ? results.hashCode() : 0);
    result = 31 * result + (next != null ? next.hashCode() : 0);
    return result;
  }
}

File: src/main/resources/vertx-sql-js/result_set.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-sql-js/result_set */
var utils = require('vertx-js/util/utils');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JResultSet = io.vertx.ext.sql.ResultSet;

/**
 Represents the results of a SQL query.
 <p>
 @class
*/
var ResultSet = function(j_val) {

  var j_resultSet = j_val;
  var that = this;

  /**

   @public

   @return {Array.<todo>}
   */
  this.getResults = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return utils.convReturnListSetJson(j_resultSet["getResults()"]());
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the results

   @public

   @return {Array.<todo>} the results
   */
  this.results = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return utils.convReturnListSetJson(j_resultSet["results()"]());
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {Array.<string>}
   */
  this.getColumnNames = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_resultSet["getColumnNames()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the column names

   @public

   @return {Array.<string>} the column names
   */
  this.columnNames = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_resultSet["columnNames()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {Array.<Object>}
   */
  this.getRows = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return utils.convReturnListSetJson(j_resultSet["getRows()"]());
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Get the rows - each row represented as a JsonObject where the keys are the column names and the values are
   the column values.
   <p>
   Beware that it's legal for a query result in SQL to contain duplicate column names, in which case one will
   overwrite the other if using this method. If that's the case use {@link ResultSet#getResults} instead.

   @public

   @return {Array.<Object>} the rows represented as JSON object instances
   */
  this.rows = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return utils.convReturnListSetJson(j_resultSet["rows()"]());
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {number}
   */
  this.getNumRows = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_resultSet["getNumRows()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Return the number of rows in the result set

   @public

   @return {number} the number of rows
   */
  this.numRows = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_resultSet["numRows()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {number}
   */
  this.getNumColumns = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_resultSet["getNumColumns()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**
   Return the number of columns in the result set

   @public

   @return {number} the number of columns
   */
  this.numColumns = function() {
    var __args = arguments;
    if (__args.length === 0) {
      return j_resultSet["numColumns()"]();
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_resultSet;
};

/**

 @memberof module:vertx-sql-js/result_set
 @param columnNames {Array.<string>} 
 @param results {Array.<todo>} 
 @return {ResultSet}
 */
ResultSet.create = function(columnNames, results) {
  var __args = arguments;
  if (__args.length === 2 && typeof __args[0] === 'object' && __args[0] instanceof Array && typeof __args[1] === 'object' && __args[1] instanceof Array) {
    return utils.convReturnVertxGen(JResultSet["create(java.util.List,java.util.List)"](columnNames, utils.convParamListJsonArray(results)), ResultSet);
  } else throw new TypeError('function invoked with invalid arguments');
};

// We export the Constructor function
module.exports = ResultSet;